### DISCUSSION TITLE
PHP 8: "PDOException: There is no active transaction"

### DISCUSSION THREAD

**Discussion by JarJak** [doctrine/doctrine2#11542] [Support Questions]
### Bug Report

I am not sure if this is bug for ORM or DBAL.

Versions:
PHP 8.2
ORM 2.19.5
DBAL 3.8.6
Symfony 8.2
Postgres 13.7

DBAL config:

```
                'connections' => [
                    'default' => [
                        'url' => '%env(resolve:DATABASE_URL)%',
                        'use_savepoints' => true,
                        'driver' => 'pdo_pgsql',
                    ],
                ],
```

#### Summary

When using transactions inside code, when exception occurs "entitymanager is closed" and transaction is rollbacked, that results in  "PDOException: There is no active transaction" error.

I have seen similar bugs reported, but they either happen with MySQL/MariaDB or in migrations. For me this happens with Postgres and in regular app code.

Even though before rollbacking, Doctrine checks if it has a flag for active transaction before sending `ROLLBACK` sql to the database, for some reason this does not always work.

Code fragment from UoW commit() method (which is always called in flush() regardless of transaction mode): https://github.com/doctrine/orm/blob/3.2.x/src/UnitOfWork.php#L439
```php
        } catch (Throwable $e) {
            // exception "entitymanager is closed" caught
            $this->em->close(); // closes em (but is already closed)

            if ($conn->isTransactionActive()) { // this for some reason returns true
                $conn->rollBack(); // this fails with "no active transaction"
            }

            $this->afterTransactionRolledBack();

            throw $e;
        }
```

#### Proposed solution

One "dirty and hacky" way to fix it would be to add another try/catch inside the code above:

```php
        } catch (Throwable $e) {
            // exception "entitymanager is closed" caught
            $this->em->close(); // closes em (but is already closed)

            if ($conn->isTransactionActive()) { // this for some reason returns true
                try {
                	$conn->rollBack(); // this fails with "no active transaction"
                } catch (\PDOException $pdoException) {
                    if ($pdoException->getMessage() !== 'There is no active transaction') {
                        // silent only this exception
                        throw $pdoException;
                    }
                }
            }

            $this->afterTransactionRolledBack();

            throw $e;
        }
```

Or to do this try/catch inside `Connection->rollBack()`

**Comment by derrabus** [comment]
First of all, it's PDO being picky about the transaction status. Postgres allows to rollback without an active transaction, but PDO throws. 🤷🏻‍♂️ 

If we want to file this as a bug, I need a reproducer for the issue. Otherwise, we can neither determine the cause of your problem nor find the right fix. Moving to support for the time being.
